object_oriented_programming.ipynb

```diff
@@ -0,0 +1,1534 @@
+   "outputs": [],
+   "source": [
+    "str(r1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "851e869e-66d0-4cd5-a8cb-b4d51b274ac1",
+   "metadata": {},
+   "source": [
+    "This is where these special methods come in. When we call `str(r1)`, Python will first look to see if our class (`Rectangle`) has a special method called `__str__`.\n",
+    "\n",
+    "If the `__str__` method is present, then Python will call it and return that value.\n",
+    "\n",
+    "There's actually another one called `__repr__` which is related, but we'll just focus on `__str__` for now."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "98c51761-ac50-4d53-a780-9332e6da3528",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "class Rectangle:\n",
+    "    def __init__(self, width, height):\n",
+    "        self.width = width\n",
+    "        self.height = height\n",
+    "        \n",
+    "    def area(self):\n",
+    "        return self.width * self.height\n",
+    "    \n",
+    "    def perimeter(self):\n",
+    "        return 2 * (self.width + self.height)\n",
+    "    \n",
+    "    def __str__(self):\n",
+    "        \"\"\"A string representation of a Rectangle\"\"\"\n",
+    "        return f'Rectangle (width={self.width}, height={self.height})'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9720291b-68dc-40ec-b692-e9b4da387904",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "r1 = Rectangle(10, 20)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b5100468-087f-4256-b82d-2f6b8778f856",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "str(r1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "422b2608-71be-4e86-b886-e08be66d42e9",
+   "metadata": {},
+   "source": [
+    "However, in Jupyter, look what happens here:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "566a823a-48f4-4a5f-a88e-c93cbf05b268",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "r1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0d793886-49fd-4300-87e5-664fc7a3eb3c",
+   "metadata": {},
+   "source": [
+    "As you can see we still get that default. That's because here Python is **not** converting `r1` to a string, but instead looking for a string *representation* of the object. It is looking for the [`__repr__` method](https://docs.python.org/3/reference/datamodel.html#object.__repr__), which is defined as\n",
+    "\n",
+    "> the “official” string representation of an object\n",
+    "\n",
+    "Ideally, the `__repr__` method should return a **valid Python expression** that can be used to recreate an instance of the object. If it's not possible, it should return a simple string. For this tutorial, we can define a `__repr__` method that behaves **identically** to the `__str__` method."
```

Comment:
Why don't  you just assign __str__ to __repr__ in the class body? no need to repeat the code and a good lesson in DRY.